File: models/Gpt-3.5-turbo/cpp/code/problems/additional_problems/Acyclic_Graph_Edges.cpp

#include <iostream>
#include <vector>

using namespace std;

void dfs(int node, const vector<vector<int>>& graph, vector<bool>& visited, ostream& out) {
    visited[node] = true;

    for (int neighbor : graph[node]) {
        if (!visited[neighbor]) {
            out << node + 1 << " " << neighbor + 1 << "\n";
            dfs(neighbor, graph, visited, out);
            out << neighbor + 1 << " " << node + 1 << "\n";
        }
    }
}

int main() {
    int n, m;
    cin >> n >> m;

    vector<vector<int>> graph(n);

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        graph[a - 1].push_back(b - 1);
        graph[b - 1].push_back(a - 1);
    }

    vector<bool> visited(n, false);

    dfs(0, graph, visited, cout);

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/additional_problems/Acyclic_Graph_Edges.py

from collections import defaultdict

def dfs(node, graph, visited, ordering):
    visited[node] = True
    
    for neighbor in graph[node]:
        if not visited[neighbor]:
            dfs(neighbor, graph, visited, ordering)
            
    ordering.insert(0, node)

def solve(n, m, edges):
    graph = defaultdict(list)
    
    for a, b in edges:
        graph[a].append(b)
        
    visited = [False] * (n+1)
    ordering = []
    
    for node in range(1, n+1):
        if not visited[node]:
            dfs(node, graph, visited, ordering)
            
    ordering.reverse()
    
    directions = []
    
    for node in ordering:
        for neighbor in graph[node]:
            directions.append((node, neighbor))
    
    return directions

# Read input
n, m = map(int, input().split())

edges = []
for _ in range(m):
    a, b = map(int, input().split())
    edges.append((a, b))

# Solve the problem
directions = solve(n, m, edges)

# Print the directions
for a, b in directions:
    print(a, b)
